lib/puppet/parser/functions/delete.rb

```diff
@@ -20,19 +20,28 @@ module Puppet::Parser::Functions
 
     delete('abracadabra', 'bra')
     Would return: 'acada'
+
+    delete(['abracadabra'], '^.*bra.*$', true)
+    Would return: []
+
+    delete(['abracadabra'], '^.*jimbob.*$', true)
+    Would return: ['abracadabra']
   EOS
   ) do |arguments|
 
-    if (arguments.size != 2) then
-      raise(Puppet::ParseError, "delete(): Wrong number of arguments "+
-                                  "given #{arguments.size} for 2.")
-    end
+    raise(Puppet::ParseError, "delete(): Wrong number of arguments "+
+                              "given #{arguments.size} for 2 or 3.") unless (2..3).include? arguments.size
 
     collection = arguments[0].dup
+    use_regex = arguments.size == 3 ? arguments[2] : false
     Array(arguments[1]).each do |item|
       case collection
         when Array, Hash
-          collection.delete_if { |coll_item| coll_item =~ %r{#{item}} }
+          if use_regex
+            collection.reject! { |coll_item| (coll_item =~ %r{#{item}}) } # make no assumptions about regex format
+          else
+            collection.reject! { |coll_item| coll_item == item }
```

Comment:
Hey @jyaworski, we can't merge this change, because right here it changes old behavior and is backwards incompatible. Given the old code if you had `delete("foobarfoo", "bar")` then the result is `foofoo`. In this version, this same code would result in `foobarfoo` since it is doing absolute compares.